telegram_houseplant_bot.py

```
@@ -0,0 +1,552 @@
+    if CONFIGS['telegram']['chat-id'] == update.message.chat_id:
+        result = send_metadata(context.user_data['plant'])
+
+        context.user_data['plant'].clear()
+        if result == 0:
+            await update.message.reply_text(
+                "You've confirmed your metadata entry, and it has been updated. "
+                "Use /results see the results."
+            )
+        else:
+            await update.message.reply_text(
+                "Your metadata was not sent; please use /update_plant to re-try."
+            )
+
+        return ConversationHandler.END
+    else:
+        await update.message.reply_text(
+                "You are you authorized to use this application."
+            )
+        return ConversationHandler.END
+
+
+def send_metadata(metadata): 
+    # 1. set up schema registry
+    sr_conf = {
+        'url': CONFIGS['schema-registry']['schema.registry.url'],
+        'basic.auth.user.info': CONFIGS['schema-registry']['basic.auth.user.info']
+    }
+    schema_registry_client = SchemaRegistryClient(sr_conf)
+
+    # 2. set up metadata producer
+    avro_serializer = AvroSerializer(
+            schema_registry_client = schema_registry_client,
+            schema_str = avro_helper.houseplant_schema,
+            to_dict = avro_helper.Houseplant.houseplant_to_dict
+    )
+
+    producer_conf = CONFIGS['kafka']
+    producer_conf['value.serializer'] = avro_serializer
+    producer = SerializingProducer(producer_conf)
+
+    # 3. send metadata message
+    try:
+        value = avro_helper.Houseplant.dict_to_houseplant(metadata)
+
+        k = str(metadata.get('plant_id'))
+        logger.info('Publishing metadata message for key ' + str(k))
+        producer.produce(METADATA_TOPIC, key=k, value=value) 
+        producer.poll()
+        producer.flush()
+    except Exception as e:
+        print(str(e))
+        logger.error('Got exception ' + str(e))
+        return 1
+
```

I would just:

```python
except Exception as e:
  logger.error("Got exception %s", e)
  raise e
```

ie. Don't swallow the exception then return 1 - instead give the caller the exception back. Often they'll get more useful information from the runtime's default handler. In fact, the only reason I'd catch the unhandlable exception is to get it into the logger.

Also, `"Got exception %s", e` is neater.

---

Ah wait, I take that back. the `1` and `0` here aren't exit codes, just return codes. Consider an enum here, instead of magic numbers?

---

I raised it anyway and addressed in the next function.